### ISSUE TITLE
Tracking issue for const `<*const T>::is_null`

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#74939]
This is tracking `#[feature(const_ptr_is_null)]`.

```rust
impl<T: ?Sized> *const T {
    pub const unsafe fn as_ref<'a>(self) -> Option<&'a T>;
    pub const fn is_null(&self) -> bool;
}

impl<T: ?Sized> *mut T {
    pub const unsafe fn as_mut<'a>(self) -> Option<&'a mut T>;
    pub const unsafe fn as_ref<'a>(self) -> Option<&'a T>;
    pub const fn is_null(&self) -> bool;
}
```

Comparing pointers in const eval is [dangerous business](https://github.com/rust-lang/rust/issues/53020).

But checking whether a pointer is the null pointer is actually completely fine, as Rust does not support items being placed at the null address. Any otherwise created null pointers are supposed to return `true` for `is_null` anyway, so that's ok. Thus, we implement `is_null` as `ptr.guaranteed_eq(ptr::null())`, which returns true if it's guaranteed that `ptr` is null, and there are no cases where it will return false where it *may be* `null`, but we don't know.

**Comment by RalfJung** [comment]
> and there are no cases where it will return false where it may be null, but we don't know.

That's not entirely correct. It will at some point be possible to create out-of-bounds raw pointers in const-eval (it might be possible already with enough hacks, I am not sure). Once that is possible, a pointer could be offset enough such that it is NULL at run-time but still non-NULL when checked during CTFE -- we cannot know if out-of-bounds pointers are NULL as that depends on their concrete base address.

**Comment by oli-obk** [comment]
So, some examples that we ran reliably check:

* `std::ptr::null::<T>().is_null()` will always be true
* `NonZeroUsize::new(x).unwrap().is_null()` will always be false
* `(&x as *const T).is_null() will always be false
* `((&x as *const T as usize - &x as *const T as usize) as *const T).is_null()` will always be true (if you get it or an equivalent to compile)

The problematic case `(&x as *const T).offset(496).is_null()` will be `false`, but at runtime could be `true`. We could set it up so it would be `true` instead, but then `is_null()` can return `true` when at runtime it would return `false`.

Another alternative is to panic if no exact solution is known (this can be done in a way that it compiles away to nothing at runtime)

Basically the options are:

* use `ptr.guaranteed_eq(std::ptr::null())`, but then you get `false` for out of bounds pointers that at runtime are null
* use `!ptr.guaranteed_ne(std::ptr::null())`, but then you get `true` for all out of bounds pointers.
* use both variants and assert that they return the same thing.

**Comment by josephlr** [comment]
> * use both variants and assert that they return the same thing.

As a user, this seems the most natural to me.
 - Integer pointer that's zero => return `true`
 - Integer pointer that's non-zero => return `false`
 - Real pointer that's in-bounds => return `false`
 - Real pointer that's out-of-bounds => panic, could be either `true`/`false` at runtime.

In general, panicking when doing something fundamentally non-const a `cosnt fn` seems fine.

**Comment by kupiakos** [comment]
Is there anything blocking stabilization of this? This also blocks [`NonNull::new`](https://github.com/rust-lang/rust/issues/93235) from stabilization.

**Comment by RalfJung** [comment]
Given in particular the use of `is_null` in `NonNull::new`/`ptr::as_ref`, I think the only possible interpretation it can have is "might be null": if we allow `NonNull::new(ptr.wrapping_sub(0x120000)).unwrap()` and later it turns out that pointer is at address 0x120000, we have a problem. Therefore, `ptr.wrapping_sub(0x120000).is_null` must be `true`.

Or of course we could panic and thus abort const-eval. We'll never panic at runtime so this may be acceptable.

@rust-lang/wg-const-eval any opinions on this?

**Comment by lunlian** [comment]
https://github.com/lgslix
https://github.com/znymo

**Comment by D0liphin** [comment]
# Stabilization Report

## Summary

[Current tracking issue](https://github.com/rust-lang/rust/issues/74939)

Pointers can be definitely null:
```rs
std::ptr::null().is_null() // always `true`
```
Or definitely not null:
```rs
(&x as *const T).is_null() // always `false` 
```
Sometimes though, pointers are "maybe null", which is what makes stabilizing
this tricky.
```rs
(&x as *const T).wrapping_offset(-0x12345).is_null() // who knows?
```
The "end-to-end behavior" **must be the same** for both the `const fn` and 
`fn`. This rules out selecting either `true` or `false` consistently for the
maybe case; it could be different at runtime.

What we're left with then is no option but to abort compilation if 
`guaranteed_eq` fails (by panicking). We actually abort compilation in a 
`const` context, so there is "no behaviour" so to speak. 

### (also) Stabilize `<*const T>.as_ref()` as a `const fn`

This is currently blocked on `const_ptr_is_null`, so follows immediately
from stabilizing it. We can't do the same for `<*mut T>.as_ref()`, that's 
blocking on `const_mut_refs`, now.

### (also) Stabilize `NonNull::new()` as a `const fn`

[Current tracking issue](https://github.com/rust-lang/rust/issues/93235)

This follows immediately from stabilizing the above, but would need a separate stabilization process, I suppose?

## Tests

- `const_nonnull_new` is tested in a `const` context [here](https://github.com/rust-lang/rust/blob/cdd0873db690cbcea1e5a011bbfc7fc94c5d5a6e/library/core/tests/ptr.rs#L278).
- `const_ptr_is_null` is tested in a `const` context [here](https://github.com/oli-obk/rust/blob/34c3c0dae5389b345e2a53991425972c02a98abc/src/test/ui/consts/ptr_is_null.rs)

**Comment by workingjubilee** [comment]
I have updated the tracking issue with all const fn listed under this feature flag.

**Comment by joshtriplett** [comment]
We talked about this in today's @rust-lang/libs-api meeting. libs-api would like to largely delegate the pointer-semantics aspects of this to lang (or opsem or whichever is appropriate). The main property we'd like to see preserved is that `is_null` should ideally not give a definitive `true` or `false` answer at compile time but then give a different answer at runtime.

**Comment by RalfJung** [comment]
> If that can only happen if you do an out-of-bounds computation using something like wrapping_offset, that seems perfect.

Yes, that is exactly the case.

> The main property we'd like to see preserved is that is_null should ideally not give a definitive true or false answer at compile time but then give a different answer at runtime.

Indeed, I think that would be a big problem. That's why I made that case into a panic.

**Comment by m-ou-se** [comment]
In general, I think it's perfectly reasonable for there to be (uncommon) situations where const eval results in a panic and runtime eval doesn't. (Unlike the other way around, or giving different return values, etc.)

**Comment by RalfJung** [comment]
[This message](https://github.com/rust-lang/rust/issues/74939#issuecomment-2336740597) has 👍 from all of wg-const eval, so what remains is the lang team.  @rust-lang/lang we'd like to make `is_null` const-callable, so [here](https://github.com/rust-lang/rust/issues/74939#issuecomment-2381628703) for the detailed report about the tricky part of this question. :)

**Comment by nikomatsakis** [comment]
My mental model for const eval is that it can sometimes fail to evaluate (panic) in value-dependent ways that are not tracked by the type system. For example, integers created by (exposed) pointers or other things may support a limited set of operations, etc, or (under nightly extensions to const generics) we may wish to distinguish `&`-references that refer to a `static` from other `&`-references. 

This seems to fit that model fairly well-- do something complex enough in a const fn, and the value you produce may not be usable in all downstream operations, even if the type system would otherwise allow it.

Is that a reasonable mental model or am I missing something?

**Comment by RalfJung** [comment]
That sounds right. Note however that there is no UB-free way to turn a pointer into an integer in const eval.

**Comment by traviscross** [comment]
@rfcbot reviewed

@rustbot labels -I-lang-nominated

**Comment by RalfJung** [comment]
@Amanieu @BurntSushi @m-ou-se @nikomatsakis @pnkfelix @tmandry FCP checkbox reminder ping :)

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by RalfJung** [comment]
Stabilization PR is up. :)
https://github.com/rust-lang/rust/pull/133116
